jooby/src/main/java/io/jooby/spi/ValueConverters.java

```diff
@@ -0,0 +1,95 @@
+package io.jooby.spi;
+
+import java.util.ServiceLoader;
+import java.util.concurrent.CopyOnWriteArrayList;
+
+import javax.annotation.Nullable;
+
+import io.jooby.TypeMismatchException;
+
+/**
+ * Contains the {@link ValueConverter}s loaded via the ServiceLoader. It is is a
+ * singleton and an instance can be retrieved with {@link #getInstance()}. The
+ * ValueConverters are stored in an ordered collection and thus resolution of
+ * type to be converted is based on the order of the converters.
+ *
+ * @author agentgt
+ *
+ */
+public final class ValueConverters {
+
+  // Allow thread safe adding of ValueConverters.
+  private final CopyOnWriteArrayList<ValueConverter> valueConverters;
```

Comment:
Mainly because it was easy for the moment.

It’s either make it so you can never change it once initialized (on demand idiom) which is painful for unit testing or use a volatile field and do all the double check locking logic. The volatile would be the whole ValueConverters reference.

Volatile field access can be pretty slow but probably faster than copyonwritearray.

Reply:
Actually the COW is probably faster than volatile. I would basically have to do what it is doing and use locks to avoid volatile field access potentially on ever request.

Of course we could not use a volatile and just allow accidental reinit each time and other odd behavior.

Reply:
I’ll look into doing it though but I’m fairly sure it will be more code as I have done it in the past with double check locking.

Reply:
I gave up and just made it volatile. The COW's internal array is volatile.